### PR TITLE
Development

### PR DESCRIPTION
**Pull request by CihatAltiparmak**

### CODE REVIEW

**Comment by CihatAltiparmak** [comment]
Hello @sjahr  and @henningkayser , could you review this PR? I need your help for this PR. I have added a documentation. I have already answered and applied necessary modifications for @sjahr 's review in #1

File: CMakeLists.txt
Comment (sjahr) [inline CMakeLists.txt]: Replace ${PROJECT_NAME}  with moveit_middleware_benchmark (also applies to comment below)

File: README.md
Comment (sjahr) [inline README.md]: Maybe you can add a small section here how to visualize the results (You can also open an issue and do that in a follow-up PR)

File: config/scenario_perception_pipeline_test_cases.yaml
Comment (sjahr) [inline config/scenario_perception_pipeline_test_cases.yaml:2]: Can you add a short comment that describes the test cases and why you selected these ones? It can also be something like `X randomly selected poses for the perception pipeline benchmark` if there is nothing special about them

File: README.md
Comment (sjahr) [inline README.md:3]: Add a small descriptive abstract (3-4 sentences about the repositories components and their purpose)
```
This repository contains ...
The perception pipeline benchmark can be used to measure ...
...
```

File: docs/how_to_install.md
Comment (sjahr) [inline docs/how_to_install.md]: Is that branch still the one to use?

File: docs/how_to_run.md
Comment (sjahr) [inline docs/how_to_run.md:5]: How to you differentiate between time required by the middleware communication layer  vs the planner/ execution algorithms?

File: docs/how_to_run.md
Comment (sjahr) [inline docs/how_to_run.md:7]: Does the selected middleware have any impact on the success / failure of the planning request? I'd assume it shouldn't and in that case these numbers might be not super relevant for the benchmark. What do you think?

File: docs/index.md
Comment (sjahr) [inline docs/index.md]: I like that! Move it directly into the README.md file. The tutorials use and index.md file because it is used to build the website but for this repos, I think it is not necessary

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (sjahr) [inline docs/scenarios/perception_pipeline_benchmark.md:36]: How relevant is it which pose I select for the middleware benchmark?

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (sjahr) [inline include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Nitpick: You could group all moveit headers together

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (sjahr) [inline include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]:
```suggestion
  /** move_group_interface_ptr to communicate with move_group_server */
```

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (sjahr) [inline include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: You can drop the comments that don't give more context like this

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (sjahr) [inline include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Nitpick: I'd put the protected class members after the public ones

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (sjahr) [inline include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:125]: What's the selection criteria?

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (sjahr) [inline include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: You can drop the `\return` comment for void functions

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (sjahr) [inline launch/scenario_perception_pipeline_benchmark.launch.py]: Do you need the db_arg?

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (sjahr) [inline launch/scenario_perception_pipeline_benchmark.launch.py:41]: Is gripper controller,yaml the right controller config file?

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (sjahr) [inline launch/scenario_perception_pipeline_benchmark.launch.py]: I think this can be removed, it is not used, an Mongo DB might not even be available in ROS 2 (this information might be  a couple of months old though :sweat_smile: )

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (sjahr) [inline launch/scenario_perception_pipeline_benchmark.launch.py]: Remove?

File: package.xml
Comment (sjahr) [inline package.xml]:
```suggestion
  <description>Middleware Benchmark Tool For MoveIt 2</description>
```

File: src/scenario_perception_pipeline_benchmark_main.cpp
Comment (sjahr) [inline src/scenario_perception_pipeline_benchmark_main.cpp:40]: Add license

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (sjahr) [inline src/scenarios/scenario_perception_pipeline.cpp]: In cases where the input is not changed you should make it const

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (sjahr) [inline src/scenarios/scenario_perception_pipeline.cpp:129]: When possible, prefere `.at(i)` over `[i]`, while it is a bit slower, you get a nicer segfault message when trying to access something out of bounds

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (sjahr) [inline src/scenarios/scenario_perception_pipeline.cpp:156]: I am confused, where is the elapsed time measured?

**Review by sjahr — COMMENTED** [review]
Well-structured code and good amount of in-code documentation, thanks!
 Is this PR supposed to measure the elapsed time as will or is it just setting up the basic infrastructure? Also I don't understand how you are differentiating between the middleware processing time and the planning time (for the middleware benchmark we only care for the first).

File: CMakeLists.txt
Comment (CihatAltiparmak) [reply CMakeLists.txt]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/0ba81e8b859e5ef794b13d768238d1c2cce5b58f commit

File: config/scenario_perception_pipeline_test_cases.yaml
Comment (CihatAltiparmak) [reply config/scenario_perception_pipeline_test_cases.yaml:2]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/3857b680b32a083d9d95825405ecf06e9973e0a8 commit

File: README.md
Comment (CihatAltiparmak) [reply README.md:3]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/fa4710c5de5217be352e51797c365f8fc1e1ae1f commit

File: docs/how_to_install.md
Comment (CihatAltiparmak) [reply docs/how_to_install.md]: No, this one was for under construction period. Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/88543a0213a9b9d3add4116ee954a7ce5518d05c commit. But it is better to work on development at the moment. Because we still haven't merged this branch(development) to main.

File: docs/index.md
Comment (CihatAltiparmak) [reply docs/index.md]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/57a5a2ce4de5ae5663a36a34fb412d9509c7e6d6 commit. I hope this one improves README.md. Btw i also added CI badge :smiley:

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (CihatAltiparmak) [reply include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/066e8dff0e01e2824773f62826e58375a38fc4d4 commit

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (CihatAltiparmak) [reply include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (CihatAltiparmak) [reply include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (CihatAltiparmak) [reply include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (CihatAltiparmak) [reply launch/scenario_perception_pipeline_benchmark.launch.py]: Removed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/caa54960dd6febab1c82f7b63b93a38616339cda commit

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (CihatAltiparmak) [reply launch/scenario_perception_pipeline_benchmark.launch.py]: Also removed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/caa54960dd6febab1c82f7b63b93a38616339cda commit

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (CihatAltiparmak) [reply launch/scenario_perception_pipeline_benchmark.launch.py]: Also removed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/caa54960dd6febab1c82f7b63b93a38616339cda commit

File: package.xml
Comment (CihatAltiparmak) [reply package.xml]: Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/fa4710c5de5217be352e51797c365f8fc1e1ae1f commit

File: src/scenario_perception_pipeline_benchmark_main.cpp
Comment (CihatAltiparmak) [reply src/scenario_perception_pipeline_benchmark_main.cpp:40]: Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/dafc402a5f4ed41de6159a366e71677f5feea573 commit

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (CihatAltiparmak) [reply src/scenarios/scenario_perception_pipeline.cpp]: I fixed it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/1acebf40d7052970148dc6bd4a9ab84a2e5fd1bb commit. But I'm wondering  whether below code part should be returned const or not. Do you have any idea @sjahr ?

https://github.com/CihatAltiparmak/moveit_middleware_benchmark/blob/caa54960dd6febab1c82f7b63b93a38616339cda/src/scenarios/scenario_perception_pipeline.cpp#L131-L139

File: README.md
Comment (CihatAltiparmak) [reply README.md]: I actually implemented visualizer script in python. It would be great to prepare visualization relavant stuff in follow-up PR. But if you wonder it, i run benchmarks separately per each middleware (rmw_zenoh, rmw_cyclonedds and rmw_fastrtps). Then i move this files to where the script is located and run scripts. Box plot is created magically after running script. Script is [here](https://gist.github.com/CihatAltiparmak/899090aca71223fbb964569f83265a49).

File: docs/how_to_run.md
Comment (CihatAltiparmak) [reply docs/how_to_run.md:5]: The elapsed time means real time in benchmark results created by google benchmark tool. It seems there is no other way to measure the time plan-execution jobs took and the time data communication jobs took. Because i just implemented benchmark in the client side (move_group_interface). Instead, i follow the strategy that 

* to measure elapsed time (wall time, real time in other meaning)  
  * send pose msg for the robot to plan
  * wait for servers response (data communication jobs + planning jobs) 
  * send plan msg for robot to execute trajectory (data communication jobs + execution jobs) 
* to do this measurements per each middleware
* compare the results

If we make the total time of planning jobs and execution jobs same for each middleware, planning jobs and execution jobs don't affect comparison of middleware. To send much goal pose (i'm speaking for perception pipeline) decreases randomness of ompl planner etc. The real aim isn't to mesaure the time, but to compare the time per each middleware.

File: docs/how_to_run.md
Comment (CihatAltiparmak) [reply docs/how_to_run.md:7]: You're right. But we are working on experimental middleware as well. I'm trying to follow each updates in rmw_zenoh. It is possible that forward updates on rmw_zenoh break the benchmark pipeline. rmw_zenoh had problem for multithreaded executor before(It seems it's fixed now but there is still no guarantee that rmw_zenoh will not break our benchmark) and this bug sometimes broke our scenario before. (communication between action_client and action_server). To sum up, i want not to remove this part because of above reasons. But if you think your idea is still better than mine, i can change this part.

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (CihatAltiparmak) [reply docs/scenarios/perception_pipeline_benchmark.md:36]: Not to select pose that planner gets hard to calculate path is better for perception pipeline. Because we are measuring for middleware, not for planner.  Should i add this part to documentation, @sjahr ?

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (CihatAltiparmak) [reply include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:125]: I'm passing parameters to benchmark node like below. User selects one of the test cases and only selected test case is benchmarked.

https://github.com/CihatAltiparmak/moveit_middleware_benchmark/blob/caa54960dd6febab1c82f7b63b93a38616339cda/launch/scenario_perception_pipeline_benchmark.launch.py#L128-L134

File: launch/scenario_perception_pipeline_benchmark.launch.py
Comment (CihatAltiparmak) [reply launch/scenario_perception_pipeline_benchmark.launch.py:41]: Honestly, i don't know, I have stolen most of it from moveit_resources repo :smile_cat: .

https://github.com/moveit/moveit_resources/blob/55d3ce0b8b95ddd224c3892ce49c37c0dc9c9361/panda_moveit_config/launch/demo.launch.py#L41-L49

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (CihatAltiparmak) [reply src/scenarios/scenario_perception_pipeline.cpp:129]: I couldn't understand where i apply this changes. @sjahr Could you explain this part in more detailed ? Btw i try to use your suggestion everywhere.

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (CihatAltiparmak) [reply src/scenarios/scenario_perception_pipeline.cpp:156]: Google benchmark library handles this stuff automatically. `st.counters` is used optionally for additional outputs like failure rate, sucess_rate. For more, you can take a cursory look at https://github.com/google/benchmark/blob/main/docs/user_guide.md#counter-reporting If you are still confused, ping me again.

File: include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp
Comment (CihatAltiparmak) [reply include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp]: Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

**Comment by CihatAltiparmak** [comment]
> Well-structured code and good amount of in-code documentation, thanks! Is this PR supposed to measure the elapsed time as will or is it just setting up the basic infrastructure?

Firstly thank you so much for your valuable review. This PR creates a json report after running launch file as indicated in this repo's documentation. But launch doesn't close itself as soon as the execution of benchmark_node finishes. You must close it (Btw i have a fix for that and i'm working this on servo_benchmarks ). For example report:

https://gist.github.com/CihatAltiparmak/ab3676b0452203f36f26f110b9a5b5fe

I'm planning to add some dockerfile for this PR (probably in another branch created from development) for your working on better

**Comment by CihatAltiparmak** [comment]
Hello @sjahr , i think i'm ready for a second review. When you have any chance etc. to take a second look at this PR, could you review new commits? Btw i'm trying not to merge other branches to development branch so that you can review easier. Additionaly, I've created a github action for creating and pushing docker image from github registry. You will also have a chance to pull this docker image approximately in this morning :) But leave this merging stuffs to me at the moment. Because for example this commit(https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/e61a02a2c64262d5b2982a47f1d4a83fd804ea21) may not solve my problem completely.

File: README.md
Comment (sjahr) [reply README.md]: Ok :+1: I think you can add it to this repository in a folder called scripts.

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (sjahr) [reply docs/scenarios/perception_pipeline_benchmark.md:36]: Yes, I think that would be nice. But you can also do that in a follow-up PR

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (sjahr) [reply src/scenarios/scenario_perception_pipeline.cpp:129]: Oh you're right, I guess I thought that config is a vector but I see that the type is `YAML::Node`.

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (sjahr) [reply src/scenarios/scenario_perception_pipeline.cpp:156]: Thanks!

**Review by sjahr — DISMISSED** [review]
@henningkayser Would you like to review it too before we merge it?

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (CihatAltiparmak) [reply src/scenarios/scenario_perception_pipeline.cpp]: I mark it as solved then.

File: config/scenario_perception_pipeline_test_cases.yaml
Comment (henningkayser) [inline config/scenario_perception_pipeline_test_cases.yaml:14]: I think it's odd to just see the w component here. Can orientation be left empty altogether and default to the identity orientation?

File: docs/how_to_install.md
Comment (henningkayser) [inline docs/how_to_install.md]: Setup and build the `moveit_middleware_benchmark` including all source dependencies. This assumes that ROS Rolling is being used, but other ROS distros should work as well.

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (henningkayser) [inline docs/scenarios/perception_pipeline_benchmark.md:3]: just link to the install instructions to make sure everyone has followed these steps

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (henningkayser) [inline docs/scenarios/perception_pipeline_benchmark.md:22]: why source rolling again? I would also assume that the workspace has been sourced initially as well

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (henningkayser) [inline docs/scenarios/perception_pipeline_benchmark.md:24]: Shouldn't this be set before initializing the zenoh router or any other ROS nodes?

Comment (CihatAltiparmak) [reply docs/scenarios/perception_pipeline_benchmark.md:24]: Fixed it this commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/33677ffc956c0d8c273715485003501b8f69a4ff) . Thanks.

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (CihatAltiparmak) [reply docs/scenarios/perception_pipeline_benchmark.md:22]: You're right. I just forgot it. Fixed in this commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/33677ffc956c0d8c273715485003501b8f69a4ff) Thanks.

File: docs/how_to_install.md
Comment (CihatAltiparmak) [reply docs/how_to_install.md]: Fixed it in this commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/f2d0ab70c4d01f813484a58c99b6960b2c3206a7) Thanks

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (CihatAltiparmak) [reply docs/scenarios/perception_pipeline_benchmark.md:3]: This commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/2b1c8b136302e678a25da2d74c329a8a8f3a63d4) should solve. Can you give a feedback again?

File: config/scenario_perception_pipeline_test_cases.yaml
Comment (CihatAltiparmak) [reply config/scenario_perception_pipeline_test_cases.yaml:14]: I couldn't undestand why you want to use orientation in default. I have run benchmark scenarios using this test case until now. I will find a more convenient test case for you. But it would be so great to firstly merge this development stuff and then apply the things in this review in follow-up PRs. Does it make sense @henningkayser ?

Comment (henningkayser) [reply config/scenario_perception_pipeline_test_cases.yaml:14]: I was only commenting on the readability. w=1 is the identity transform assuming it is a quaternion and the other components are set to 0. It would be easier to read this file if all components have to be specified (of course this would bloat a lot of redundant content), or to leave the orientation out and implicitly default to the identity orientation. I agree that this should not block merging the PR.

**Review by henningkayser — APPROVED** [review]

File: config/scenario_perception_pipeline_test_cases.yaml
Comment (CihatAltiparmak) [reply config/scenario_perception_pipeline_test_cases.yaml:14]: I will add the other components in follow-up PR.

File: src/scenarios/scenario_perception_pipeline.cpp
Comment (CihatAltiparmak) [reply src/scenarios/scenario_perception_pipeline.cpp:129]: Let's do this one as resolved. Maybe we can take a look at it later together with @henningkayser

File: docs/scenarios/perception_pipeline_benchmark.md
Comment (CihatAltiparmak) [reply docs/scenarios/perception_pipeline_benchmark.md:36]: Let's do this one as resolved. Maybe we can take a look at it later together with @henningkayser

**Pull request merged**
